File: Smart-Bus/Smart-Bus/Classes/Main/ZoneControl/setZones/AllSystems/Controller/SHAreaSettingViewController.m
//
//  SHAreaSettingViewController.m
//  Smart-Bus
//
//  Created by Mark Liu on 2017/6/15.
//  Copyright © 2017年 Mark Liu. All rights reserved.
//

#import "SHAreaSettingViewController.h"

/// 系统图片最大ID
NSUInteger maxIconIDForDataBase = 10;

/// 重用cell
static NSString *systemCellReusableIdentifier = @"SHSetSystemViewCell";

@interface SHAreaSettingViewController () <UITableViewDelegate, UITableViewDataSource, UITextFieldDelegate, UINavigationControllerDelegate, UIImagePickerControllerDelegate>

/// 区域高度
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *zoneViewHeightConstraint;

/// 按钮宽度
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *iconButtonWidthConstraint;

/// 按钮高度
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *iconButtonHeightConstraint;

/// 区域
@property (weak, nonatomic) IBOutlet UIButton *iconButton;

/// 名称
@property (weak, nonatomic) IBOutlet UITextField *nameTextField;

/// 设置列表
@property (weak, nonatomic) IBOutlet UITableView *deviceListView;

/// 设置名称
@property (strong, nonatomic) NSArray *deviceNames;

/// 区域中包含的所有开启功能的系统设备
@property (nonatomic, strong) NSMutableArray *allSystems;

@end

@implementation SHAreaSettingViewController


// MARK: - 照片的处理

/// 图片按钮点击
- (IBAction)iconButtonClick {
    
    TYCustomAlertView *alertView = [TYCustomAlertView alertViewWithTitle:nil message:@"Change zone picture?" isCustom:YES];
    
    // 图片库中获取
    [alertView addAction:[TYAlertAction actionWithTitle:@"Library" style:TYAlertActionStyleDefault handler:^(TYAlertAction *action) {
        
        SHZoneIconViewController *zoneIconViewController = [[SHZoneIconViewController alloc] init];
        
        zoneIconViewController.selectImage =  ^(SHIcon *icon) {
            
            if ([icon.iconName isEqualToString: self.currentZone.zoneIconName]) {
                return ;
            }
            self.currentZone.zoneIconName = icon.iconName;
            
            UIImage *zoneImage;
            if (icon.iconID > maxIconIDForDataBase) {
                
                zoneImage = [UIImage imageWithData:icon.iconData];
                
            } else {
                
                zoneImage = [UIImage imageNamed:self.currentZone.zoneIconName];
            }
            
            [self.iconButton setImage:[self.iconButton.imageView circleImageWithImage:zoneImage borderWidth:0 borderColor: nil] forState:UIControlStateNormal];
            
            [SHSQLiteManager.shared updateZone:self.currentZone];
        };
        
        SHNavigationController *navigationController = [[SHNavigationController alloc] initWithRootViewController:zoneIconViewController];
        
        [self presentViewController:navigationController animated:YES completion:nil];
        
    }]];
    
    // 相册中获取
    [alertView addAction:[TYAlertAction actionWithTitle:@"Photos" style:TYAlertActionStyleDefault handler:^(TYAlertAction *action) {
        
        if (![UIImagePickerController isSourceTypeAvailable:UIImagePickerControllerSourceTypePhotoLibrary]) {
            
            return;
        }
        
        UIImagePickerController *picker = [[UIImagePickerController alloc] init];
        picker.sourceType = UIImagePickerControllerSourceTypePhotoLibrary;
        picker.delegate = self;
        
        [self presentViewController:picker animated:YES completion:nil];
        
    }]];
    
    // 相机中获取
    [alertView addAction:[TYAlertAction actionWithTitle:@"Camera" style:TYAlertActionStyleDefault handler:^(TYAlertAction *action) {
        
        if (![UIImagePickerController isSourceTypeAvailable:UIImagePickerControllerSourceTypeCamera]) {
            
            return;
        }
        
        UIImagePickerController *picker = [[UIImagePickerController alloc] init];
        
        picker.sourceType = UIImagePickerControllerSourceTypeCamera;
        
        picker.delegate = self;
        
        [self presentViewController:picker animated:YES completion:nil];
        
    }]];
    
    // 取消
    [alertView addAction:[TYAlertAction actionWithTitle:@"Cancel" style:TYAlertActionStyleCancel handler:nil]];
    
    TYAlertController *alertController = [TYAlertController alertControllerWithAlertView:alertView preferredStyle:TYAlertControllerStyleAlert transitionAnimation:TYAlertTransitionAnimationDropDown];
    
    alertController.backgoundTapDismissEnable = YES;
    
    [self presentViewController:alertController animated:YES completion:nil];
}

// MARK: - 照片的代理

/// 取消操作
- (void)imagePickerControllerDidCancel:(UIImagePickerController *)picker
{
    [picker dismissViewControllerAnimated:YES completion:nil  ];
}

/// 获得照片
- (void)imagePickerController:(UIImagePickerController *)picker didFinishPickingMediaWithInfo:(NSDictionary<NSString *,id> *)info {
    
    [picker dismissViewControllerAnimated:YES completion:nil];
    
    // 尺寸
    CGSize iconSize =
        CGSizeMake(navigationBarHeight * 2,
                   navigationBarHeight * 2
    );
    
    UIImage *sourceImage =
    [self.iconButton.imageView circleImageWithImage:[self.iconButton.imageView reSizeImage:info[UIImagePickerControllerOriginalImage] toSize: iconSize] borderWidth:3.0 borderColor: [UIColor colorWithHex:0x7b7778 alpha:1.0]];
    
    
    
    // 如果是相机，保存到相册中去
    if (picker.sourceType == UIImagePickerControllerSourceTypeCamera) {
        
        UIImageWriteToSavedPhotosAlbum(sourceImage, self, nil, nil);
    }
    
    // ===========图片需要保存在沙盒中================
    
    SHIcon *icon = [[SHIcon alloc] init];
    icon.iconID =
        SHSQLiteManager.shared.getMaxIconID + 1;
    icon.iconName = [NSString stringWithFormat:@"icon_%tu", icon.iconID];
    icon.iconData = UIImagePNGRepresentation(sourceImage);
  
    [SHSQLiteManager.shared insertIcon:icon];
    
    self.currentZone.zoneIconName = icon.iconName;
    
    [SHSQLiteManager.shared updateZone:self.currentZone];
    
    [self.iconButton setImage:sourceImage forState:UIControlStateNormal];
}

// MARK: - textField代理

/// 开始编辑
- (void)textFieldDidBeginEditing:(UITextField *)textField {
    
    textField.borderStyle = UITextBorderStyleRoundedRect;
    textField.backgroundColor = [UIColor whiteColor];
    textField.textColor = [UIColor colorWithWhite:0.3 alpha:1.0];
}

/// 结束编辑
- (void)textFieldDidEndEditing:(UITextField *)textField {
    
    textField.borderStyle = UITextBorderStyleNone;
    textField.backgroundColor = [UIColor clearColor];
    textField.textColor = [UIColor whiteColor];
    
    if (!textField.text || !textField.text.length || [textField.text isEqualToString:self.currentZone.zoneName]) {
        textField.text = self.currentZone.zoneName;
        return;
    }
    
    self.currentZone.zoneName = textField.text;
  
    [SHSQLiteManager.shared updateZone:self.currentZone];
}

/// 退出
- (BOOL)textFieldShouldReturn:(UITextField *)textField {
    
    [textField resignFirstResponder];
    
    return YES;
}

// MARK: - 数据源

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    
    return self.deviceNames.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    SHSetSystemViewCell *cell = [tableView dequeueReusableCellWithIdentifier:systemCellReusableIdentifier forIndexPath:indexPath];
    
    cell.deviceName = self.deviceNames[indexPath.row];
    
    cell.hasDevice = [self.allSystems containsObject:@(indexPath.row + 1)];
    
    cell.choiceDevice = ^(BOOL hasChoice){
        
        NSNumber *deviceSystemID = @(indexPath.row + 1);
        
        if ([self.allSystems containsObject:deviceSystemID] && !hasChoice) {
            
            [self.allSystems removeObject:deviceSystemID];
            
        } else if (hasChoice && ![self.allSystems containsObject:deviceSystemID]){
            if (hasChoice) {
                [self.allSystems addObject:deviceSystemID];
            }
        }
    };
    
    return cell;
}


// MARK: - 视图加载

- (void)viewDidLayoutSubviews {
    
    [super viewDidLayoutSubviews];
    
    if ([UIDevice is_iPad]) {
        
        self.zoneViewHeightConstraint.constant = navigationBarHeight * 2 + statusBarHeight;
        
        self.iconButtonWidthConstraint.constant = navigationBarHeight * 2;
        self.iconButtonHeightConstraint.constant = navigationBarHeight * 2;
    }
}

/// 保存系统设备ID
- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
 
    [SHSQLiteManager.shared saveSystemIDs:self.allSystems zoneID:self.currentZone.zoneID];
    
    [self.view endEditing:YES];
}

- (void)viewWillAppear:(BOOL)animated {
    
    [super viewWillAppear:animated];
    
    NSArray *types =
        [SHSQLiteManager.shared getSystemIDs:self.currentZone.zoneID];
    
    self.allSystems = [NSMutableArray arrayWithArray:types];
  
    
    [self.deviceListView reloadData];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.navigationItem.title = @"System Setting";
    
    self.deviceNames = [SHSQLiteManager.shared getSystemNames];
    
    [self.deviceListView registerNib:[UINib nibWithNibName:systemCellReusableIdentifier bundle:nil] forCellReuseIdentifier:systemCellReusableIdentifier];
    
    self.deviceListView.rowHeight = [SHSetSystemViewCell rowHeight];
    
    UIImage *image = [UIImage imageNamed:self.currentZone.zoneIconName];
    
    if (!image) {
        
        SHIcon *icon =
            [SHSQLiteManager.shared
                getIcon:self.currentZone.zoneIconName];
        
        image = [UIImage imageWithData:icon.iconData];
    }
    
    [self.iconButton setImage:image forState:UIControlStateNormal];
    
    self.nameTextField.text = self.currentZone.zoneName;
    
    if ([UIDevice is_iPad]) {
        
        self.nameTextField.font = [UIView suitFontForPad];
    }
}

// MARK: - getter && setter

/// 所有的系统
- (NSMutableArray *)allSystems {
    
    if (!_allSystems) {
        _allSystems = [NSMutableArray array];
    }
    return _allSystems;
}

@end

File: Smart-Bus/Smart-Bus/Classes/Main/ZoneControl/setZones/AllSystems/Controller/SHAreaSettingViewController.swift
//
//  SHAreaSettingViewController.swift
//  Smart-Bus
//
//  Created by Mark Liu on 2017/6/15.
//  Copyright © 2017年 Mark Liu. All rights reserved.
//

import UIKit

/// cell 重用标示
private let systemCellReusableIdentifier =
    "SHSetSystemViewCell"

class SHAreaSettingViewController: SHViewController {
    
    /// 设置区域
    var region: SHRegion? {
        
        didSet {
            
            isSetupZone = region == nil
        }
    }
    
    /// 当前区域
    var currentZone: SHZone? {
        
        didSet {
            
            isSetupZone = currentZone != nil
        }
    }
    
    /// 是否是设置Zone
    private var isSetupZone = true
    
    /// 设置名称
    private lazy var deviceNames =
        SHSQLiteManager.shared.getSystemNames()
    
    /// 区域中包含的所有开启功能的系统设备
    private lazy var allSystems = [UInt]()
    
    /// 区域高度
    @IBOutlet weak var zoneViewHeightConstraint: NSLayoutConstraint!
    
    /// 按钮宽度
    @IBOutlet weak var iconButtonWidthConstraint: NSLayoutConstraint!
    
    /// 按钮高度
    @IBOutlet weak var iconButtonHeightConstraint: NSLayoutConstraint!
    
    /// 区域
    @IBOutlet weak var iconButton: UIButton!
    
    /// 名称
    @IBOutlet weak var nameTextField: UITextField!
    
    /// 设置列表
    @IBOutlet weak var deviceListView: UITableView!
}

// MARK: - 图片切换问题
extension SHAreaSettingViewController : UINavigationControllerDelegate, UIImagePickerControllerDelegate {
    
    /// 图片按钮点击
    @IBAction func iconButtonClick() {
        
        let alertView =
            TYCustomAlertView(
                title: nil,
                message: "Change picture?",
                isCustom: true
        )
        
        // 从图片库
        let libraryAction = TYAlertAction(title: "Library", style: .default) { (action) in
            
            let zoneIconViewController = SHZoneIconViewController()
            
            zoneIconViewController.selectImage =
                { (icon: SHIcon) -> () in
                    
                    var image: UIImage?
                    
                    if icon.iconID > maxIconIDForDataBase {
                        
                        if icon.iconData != nil {
                            
                            image =
                                UIImage(data: icon.iconData!)
                        }
                        
                    } else {
                        
                        if icon.iconName != nil {
                            
                            image = UIImage(named: icon.iconName!)
                        }
                    }
                    
                    if image == nil ||
                        icon.iconName == nil {
                        
                        return
                    }
                    
                    self.iconButton.setImage(
                        image,
                        for: .normal
                    )
                    
                    // 设置区域
                    if self.isSetupZone {
                        
                        self.currentZone!.zoneIconName = icon.iconName!
                        
                        _ = SHSQLiteManager.shared.updateZone(
                            self.currentZone!
                        )
                    
                    // 设置分区
                    } else {
                        
                        self.region!.regionIconName = icon.iconName!
                        
                        _ = SHSQLiteManager.shared.updateRegion(
                            self.region!
                        )
                    }
                    
            }
            
            let navigationIconViewController =
                SHNavigationController(
                    rootViewController:
                    zoneIconViewController
            )
            
            self.present(navigationIconViewController,
                         animated: true,
                         completion: nil
            )
        }
        
        alertView?.add(libraryAction)
        
        // 从相册
        let photoAction = TYAlertAction(title: "Photos", style: .default) { (action) in
            
            if !UIImagePickerController.isSourceTypeAvailable(.photoLibrary) {
                return
            }
            
            let picker = UIImagePickerController()
            picker.sourceType = .photoLibrary
            picker.delegate = self
            
            self.present(picker,
                         animated: true,
                         completion: nil
            )
        }
        
        alertView?.add(photoAction)
        
        
        // 从相机
        let cameraAction = TYAlertAction(title: "Camera", style: .default) { (action) in
            
            if !UIImagePickerController.isSourceTypeAvailable(.camera) {
                return
            }
            
            let picker = UIImagePickerController()
            picker.sourceType = .camera
            picker.delegate = self
            
            self.present(picker,
                         animated: true,
                         completion: nil
            )
        }
        
        alertView?.add(cameraAction)
        
        // 取消
        let cancelAction =
            TYAlertAction(
                title: "Cancel",
                style: .cancel,
                handler: nil
        )
        
        alertView?.add(cancelAction)
        
        
        let alertController =
            TYAlertController(
                alert: alertView,
                preferredStyle: .alert,
                transitionAnimation: .dropDown
        )
        
        alertController?.backgoundTapDismissEnable = true
        
        present(alertController!,
                animated: true,
                completion: nil
        )
    }
    
    // MARK: - 代理实现部分
    
    /// 取消操作
    func imagePickerControllerDidCancel(_ picker: UIImagePickerController) {
        picker.dismiss(animated: true)
    }
    
    /// 获得照片
    func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey : Any]) {
        
        picker.dismiss(animated: true)
        
        // 尺寸
        let iconSize =
            CGSize(width: navigationBarHeight * 2,
                   height: navigationBarHeight * 2)
        
        let borderColor =
            UIColor(hex: 0x7b7778,
                    alpha: 1.0
        )
        
        // 图片数据
        guard let sourceImage = info[UIImagePickerController.InfoKey.originalImage] as? UIImage,
            
            let image =
            iconButton.imageView?.reSizeImage(sourceImage, to: iconSize),
            
            let sectionImage =
            iconButton.imageView?.circleImage(
                with: image,
                borderWidth: 3.0,
                borderColor: borderColor
            )
            
            else {
                
                return
        }
        
        // 保存到相册
        if picker.sourceType == .camera {
            
            UIImageWriteToSavedPhotosAlbum(
                sectionImage, self, nil, nil
            )
        }
        
        iconButton.setImage(sectionImage, for: .normal)
        
        // 保存到数据库
        let icon = SHIcon()
        icon.iconID =
            SHSQLiteManager.shared.getMaxIconID() + 1
        
        let iconName = "icon_\(icon.iconID)"
        icon.iconName = iconName
        
        icon.iconData =
            sectionImage.pngData()
        
        _ = SHSQLiteManager.shared.insertIcon(icon)
        
        // 区域还是分组
        if isSetupZone {
            
            currentZone!.zoneIconName = iconName
            
            _ = SHSQLiteManager.shared.updateZone(currentZone!)
        
        } else {
            
            region!.regionIconName = iconName
            
            _ = SHSQLiteManager.shared.updateRegion(region!)
        }
        
    }
}



// MARK: - UITextFieldDelegate
extension SHAreaSettingViewController: UITextFieldDelegate {
    
    /// 开始编辑
    func textFieldDidBeginEditing(_ textField: UITextField) {
        
        textField.borderStyle = .roundedRect
        textField.backgroundColor = UIColor.white
        textField.textColor = UIColor(white: 0.3, alpha: 1.0)
    }
    
    func textFieldDidEndEditing(_ textField: UITextField) {
        
        textField.borderStyle = .none
        textField.backgroundColor = UIColor.clear
        textField.textColor = UIColor.white
        
        
        guard let name = textField.text else {
            return
        }
        
        if isSetupZone {
            
            currentZone?.zoneName = name
            _ = SHSQLiteManager.shared.updateZone(currentZone!)
            
        } else {
            
            region?.regionName = name
            _ = SHSQLiteManager.shared.updateRegion(region!)
        }
        
    }
    
    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        
        textField.resignFirstResponder()
        
        return true
    }
}


// MARK: - UITableViewDataSource
extension SHAreaSettingViewController: UITableViewDataSource {
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        
        return deviceNames.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        
        let cell =
            tableView.dequeueReusableCell(
                withIdentifier: systemCellReusableIdentifier,
                for: indexPath
                ) as! SHSetSystemViewCell
        
        cell.deviceName = deviceNames[indexPath.row]
        
        cell.hasDevice = allSystems.contains(UInt(indexPath.row + 1))
        
        cell.choiceDevice = { (isHaveDevice: Bool) -> Void in
            
            let systemID = UInt(indexPath.row + 1)
            
            if self.allSystems.contains(systemID) &&
                !isHaveDevice {
                
                for type in self.allSystems.enumerated() {
                    
                    if type.element == systemID {
                        
                        self.allSystems.remove(at: type.offset)
                        
                        break
                    }
                }
                
            } else if isHaveDevice &&
                !self.allSystems.contains(systemID) {
                
                self.allSystems.append(systemID)
            }
        }
        
        return cell
    }
}


extension SHAreaSettingViewController {
    
    /// 保存配置好的数据
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        
        if isSetupZone {
        
            _ = SHSQLiteManager.shared.saveSystemIDs(
                allSystems,
                zoneID: currentZone!.zoneID
            )
        }
        
        view.endEditing(true)
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        if isSetupZone {
            allSystems =
                SHSQLiteManager.shared.getSystemIDs(
                    currentZone?.zoneID ?? 0)
            
            deviceListView.reloadData()
        }
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        navigationItem.title = "Area Setting"
        
        deviceListView.isHidden = !isSetupZone
        
        deviceListView.rowHeight =
            SHSetSystemViewCell.rowHeight
        
        deviceListView.register(
            UINib(nibName: systemCellReusableIdentifier,
                  bundle: nil),
            forCellReuseIdentifier:
            systemCellReusableIdentifier
        )
        
        
        // 名称
        nameTextField.text =
            isSetupZone ? currentZone?.zoneName : region?.regionName
        
        if UIDevice.is_iPad() {
            
            nameTextField.font = UIView.suitFontForPad()
        }
        
        // 图片
        let areaIconName =
            isSetupZone ? currentZone?.zoneIconName :
                region?.regionIconName
        
        guard let iconName = areaIconName,
                let icon =
                    SHSQLiteManager.shared.getIcon(iconName),
                let image = (icon.iconData == nil) ?
                    UIImage(named: iconName) :
                    UIImage(data: icon.iconData!) else {
            
            let defaultImage =
                isSetupZone ?
                    UIImage(named: "Ceooffice") :
                    UIImage(named: "regionIcon")
            
            iconButton.setImage(defaultImage,
                                for: .normal
            )
            
            return
        }
        
        iconButton.setImage(image, for: .normal)
    }
    
    /// 布局
    override func viewDidLayoutSubviews() {
        
        super.viewDidLayoutSubviews()
        
        if UIDevice.is_iPad() {
            
            zoneViewHeightConstraint.constant =
                navigationBarHeight * 2 + statusBarHeight
            
            iconButtonWidthConstraint.constant =
                navigationBarHeight * 2
            
            iconButtonHeightConstraint.constant =
                navigationBarHeight * 2
        }
    }
}
